KAFKA-18058: Share group state record pruning impl.

smjn opened the pull request:
### What
* In this PR, we've added a class `ShareCoordinatorOffsetsManager`, which tracks the last redundant offset for each share group state topic partition. We have also added a periodic timer job in `ShareCoordinatorService` which queries for the redundant offset at regular intervals and if a valid value is found, issues the `deleteRecords` call to the `ReplicaManager` via the `PartitionWriter`. In this way the size of the partitions is kept manageable.
* Introduced new config `share.coordinator.state.topic.prune.interval.ms` (default 5 mins).

### Why
* Currently, the `SharePartition` invokes the `DefaultStatePersister` to write data into the internal share group topic `__share_group_state`. This topic is not eligible for compaction.
* If the kafka cluster is run for a long period of time, the topic partitions in `__share_group_state` will be populated with gigantic amount of records.
* In the above scenario if a broker/cluster restarts, all the records will need to be replayed on the leader `ShareCoordinatorShard` resulting in extensive latency during startup.
* One observation is that the records follow a periodic snapshot approach. A snapshot record gets created and then a few incremental updates and after a threshold number,  a snapshot again for each share partition key.
* This implies that older records become redundant after a while since they've been superseded by new snapshots. This provides us an opportunity to relieve some pressure from the partitions.

### Testing
* Added appropriate tests for `CoordinatorPartitionWriter`, `ShareCoordinatorOffsetsManager`, `ShareCoordinatorService` and `ShareCoordinatorShard`.
* Extensive manul testing

### Sample o/p
**Broker logs**
```
[2024-11-29 15:21:38,289] INFO [UnifiedLog partition=__share_group_state-3, dir=/tmp/kraft-combined-logs] Incremented log start offset to 10 due to client delete records request (kafka.log.UnifiedLog)
```

**Records before prune**
```
{"key":{"version":0,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":0,"stateEpoch":0,"leaderEpoch":0,"startOffset":5,"stateBatches":[{"firstOffset":5,"lastOffset":6,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":0,"leaderEpoch":0,"startOffset":7,"stateBatches":[{"firstOffset":7,"lastOffset":8,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":0,"leaderEpoch":0,"startOffset":9,"stateBatches":[{"firstOffset":9,"lastOffset":9,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":0,"leaderEpoch":0,"startOffset":10,"stateBatches":[{"firstOffset":10,"lastOffset":11,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":0,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":1,"stateEpoch":0,"leaderEpoch":0,"startOffset":12,"stateBatches":[{"firstOffset":12,"lastOffset":14,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":1,"leaderEpoch":0,"startOffset":15,"stateBatches":[{"firstOffset":15,"lastOffset":16,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":1,"leaderEpoch":0,"startOffset":17,"stateBatches":[{"firstOffset":17,"lastOffset":19,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":0,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":2,"stateEpoch":0,"leaderEpoch":0,"startOffset":20,"stateBatches":[{"firstOffset":20,"lastOffset":21,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":2,"leaderEpoch":0,"startOffset":22,"stateBatches":[{"firstOffset":22,"lastOffset":27,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":1,"data":{"snapshotEpoch":2,"leaderEpoch":0,"startOffset":28,"stateBatches":[{"firstOffset":28,"lastOffset":28,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":0,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":3,"stateEpoch":0,"leaderEpoch":0,"startOffset":29,"stateBatches":[{"firstOffset":29,"lastOffset":29,"deliveryState":2,"deliveryCount":1}]}}}
```

**Records after prune**
```
{"key":{"version":0,"data":{"groupId":"gs1","topicId":"o_eD4bB7RNauIV7oPkNRCA","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":3,"stateEpoch":0,"leaderEpoch":0,"startOffset":29,"stateBatches":[{"firstOffset":29,"lastOffset":29,"deliveryState":2,"deliveryCount":1}]}}}
```

smjn committed:
Prune records initial commit.

smjn committed:
added offsets manager class.

dajac commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1175,8 +1175,8 @@ class ReplicaManager(val config: KafkaConfig,
     offsetPerPartition.map { case (topicPartition, requestedOffset) =>
-      // reject delete records operation on internal topics
-      if (Topic.isInternal(topicPartition.topic)) {
+      // reject delete records operation on internal topics except for allow listed ones
+      if (Topic.isInternal(topicPartition.topic) && !config.internalTopicsRecordDeleteAllowList.contains(topicPartition.topic)) {
```
In my opinion, this is the wrong approach because it will allow users of the cluster to delete too. When I suggested it, I thought that we would add a boolean to the method, e.g. allowInternalTopics, and use it from the component doing the deletion.

smjn replied:
Understood, will rectify

smjn committed:
replace config with boolean flag arg.

smjn committed:
add deleteRecords method to partition writer.

smjn committed:
add plumbing in share coordinator for purging.

AndrewJSchofield commented on server/src/main/java/org/apache/kafka/server/config/ShareCoordinatorConfig.java:
```diff
@@ -71,6 +72,10 @@ public class ShareCoordinatorConfig {
     public static final int APPEND_LINGER_MS_DEFAULT = 10;
     public static final String APPEND_LINGER_MS_DOC = "The duration in milliseconds that the share coordinator will wait for writes to accumulate before flushing them to disk.";
 
+    public static final String STATE_TOPIC_PRUNE_INTERVAL = "share.coordinator.state.topic.prune.interval.ms";
```
This should be `STATE_TOPIC_PRUNE_INTERVAL_MS_CONFIG` I think. All three of these should have `_MS_` as part of the name.

AndrewJSchofield commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1172,11 +1172,11 @@ class ReplicaManager(val config: KafkaConfig,
     offsetPerPartition.map { case (topicPartition, requestedOffset) =>
-      // reject delete records operation on internal topics
-      if (Topic.isInternal(topicPartition.topic)) {
+      // reject delete records operation for internal topics if allowInternalTopicDeletion is false
```
probably "unless allowInternalTopicDeletion is true" is clearer.

AndrewJSchofield commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/PartitionWriter.java:
```diff
@@ -107,4 +107,16 @@ CompletableFuture<VerificationGuard> maybeStartTransactionVerification(
+    /**
+     * Delete records from a topic partition until specified offset
+     * @param tp                The partition to delete records from
+     * @param deleteUntilOffset Offset to delete until, starting from the beginning
```
I suggest `deleteBeforeOffset` rather than `deleteUntilOffset`. The latter is clearly non-inclusive, while the latter is a bit more ambiguous. I think the effect we want here is that if I provide offset 10, then offsets up to and including 9 may be deleted, but not 10.

AndrewJSchofield commented on coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/PartitionWriter.java:
```diff
@@ -107,4 +107,16 @@ CompletableFuture<VerificationGuard> maybeStartTransactionVerification(
+    void deleteRecords(
+        TopicPartition tp,
+        long deleteUntilOffset,
+        boolean allowInternalTopicDeletion
```
This parameter is missing from the javadoc.

AndrewJSchofield commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +249,43 @@ public void startup(
+        timer.add(new TimerTask(config.shareCoordinatorTopicPruneIntervalMs()) {
+            @Override
+            public void run() {
+                for (int i = 0; i < numPartitions; i++) {
```
Would It not be the case that each shard should set up the pruning for its owned partitions? Shouldn't the pruning stop when a shard loses leadership of a partition?

smjn replied:
No - the shard (state machine) does not maintain that mapping. The information is maintained by the runtime which calls the appropriate shards, based on the context. This information is not exposed outside. We need access to https://github.com/apache/kafka/blob/trunk/coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorRuntime.java#L1877 to expose this information. Even then the shard cannot do this.

The runtime is encapsulated in the ShareCoordinatorService and only it can issue calls to the runtime. The Shard only serves to provide data related to partitions.

Using the loop approach - for a specific internal topic-partition only the correct Shard will honour the request and the others will fail silently due to NOT_COORDINATOR.

Flow is 
```
                               ShareCoordinatorShard.callback
                                                     |
                                                     |  
                      add task      task with correct shard         
ShareCoordinatorService ---->   Runtime -----> ==================== ----> EventProcessor
                                   |                   QUEUE
                                   |
                             obtain shard from TP in task

```

AndrewJSchofield replied:
Thanks for the explanation.

dajac replied:
I wonder if we should expose a method to get the list of active state machines/shards from the runtime. This would allow you to just iterate on it instead of having to list all the possibilities.

AndrewJSchofield replied:
Yes, I've been thinking about this. I'm not entirely comfortable with every broker starting a timer for every partition. I know it's harmless, but it's not exactly elegant.

smjn replied:
@dajac we don't need the coordinators, but the list of topic partitions whose coordinators are ACTIVE at the point of execution of the timer job.

CoordinatorRuntime
```
    public List<TopicPartition> activeTopicPartitions() {
        return coordinators.entrySet().stream()
            .filter(entry -> entry.getValue().state.equals(CoordinatorState.ACTIVE))
            .map(Map.Entry::getKey)
            .toList();
    }
```


Caller:
```
activeTopicPartitions.forEach(tp -> runtime.scheudle(tp, ...)
```

smjn replied:
@AndrewJSchofield @dajac removed the loop in favor of active tps

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. I have done a partial review and left some comments.

smjn committed:
review comments

smjn committed:
minor bug fix.

smjn committed:
shard unit tests.

smjn committed:
partition writer unit tests.

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +249,45 @@ public void startup(
+            "write-state-record-prune",
+            tp,
+            Duration.ofMillis(config.shareCoordinatorWriteTimeoutMs()),
+            ShareCoordinatorShard::lastRedundantOffset
+        ).whenComplete((result, exception) -> {
+            if (exception != null) {
+                log.error("Last redundant offset lookup threw an error.", exception);
+                return;
+            }
```
Here, I suggest to handle the known errors such as NOT_COORDINATOR, COORDINATOR_LOADING, etc. As you optimistically send the write operations to all the possible shards, you will get many "unknown" ones.

smjn replied:
I will remove the logging for above errors - they will happen in every case. Since they timer job is periodic, we do not need any special handling anyway.

dajac commented on core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala:
```diff
@@ -165,4 +166,21 @@ class CoordinatorPartitionWriter(
+      timeout = 0L,
+      offsetPerPartition = Map(tp -> deleteBeforeOffset),
+      responseCallback = results => deleteResults = results,
+      allowInternalTopicDeletion
```
I wonder if we could just set it to `true` here as this class is only used by the coordinator.

dajac commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManagerTest.java:
```diff
@@ -0,0 +1,206 @@
+        assertEquals(1, manager.curState().size());
+        verify(manager, times(5)).purge();
+    }
+}
```
nit: We usually have an empty line at the end.

smjn committed:
added share coord service unit tests.

smjn committed:
final touches, incorporated comments.

smjn committed:
Fixed documentation.

smjn committed:
added more unit tests, constraint to exception logging.

smjn committed:
added runtime method to fetch active tps.

smjn committed:
logging improvements.

dajac commented on core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala:
```diff
@@ -165,4 +166,21 @@ class CoordinatorPartitionWriter(
+  override def deleteRecords(tp: TopicPartition, deleteBeforeOffset: Long): Unit = {
+    var deleteResults: Map[TopicPartition, DeleteRecordsPartitionResult] = Map.empty
+    replicaManager.deleteRecords(
+      timeout = 0L,
```
I have doubts about using timeout=0 here. My understanding is that the it will delete records from the local logs but it wont wait on the replication to acknowledges the deletion from the replicas. Hence, I suppose that it will always throw an error. I suppose that we don't really care about the replication in this case but it may spam the logs with unnecessary errors.

smjn replied:
This is the same as append records https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala#L144

will add 30 second timeout nevertheless

dajac replied:
You cannot compare with append records in this case because they don't work the sam way. In append records, the key is that we use `requiredAcks = 1` which means that the operation does not go to the purgatory at all. In the delete record case, it goes to the purgatory in all cases.

With the timeout, I believe that the following code will actually throw a `IllegalStateException` because the operation in the purgatory won't be completed when you reach it.

```
    val partitionResult = deleteResults.getOrElse(tp,
      throw new IllegalStateException(s"Delete status $deleteResults should have partition $tp."))
```

Could you please double check?

One option may be to make delete records return a future and complete it when the callback returns. This would also allow you to not trigger the next purge while the current one is not completed.

dajac commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1369,9 +1369,10 @@ class ReplicaManager(val config: KafkaConfig,
 
   def deleteRecords(timeout: Long,
```
Let's add a unit test for this new flag in ReplicaManagerTest.

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +249,60 @@ public void startup(
+                    try {
+                        writer.deleteRecords(tp, off);
+                    } catch (Exception e) {
+                        log.error("Failed to delete records in {} till offset {}.", tp, off, e);
```
I believe this will be printed all the time.

smjn replied:
will change to debug

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java:
```diff
@@ -219,6 +232,8 @@ private void handleShareSnapshot(ShareSnapshotKey key, ShareSnapshotValue value)
             }
         }
+
+        offsetsManager.updateState(mapKey, offset);
```
My understanding is that you don't use a timeline data structure in the offset manager. The risk is that a snapshot maybe replayed and hence updates the offset here. However, at this point, we don't know whether the snapshot will be durably stored. Let's imagine the worst case, the write of the snapshot fails for whatever reason. In this case, the snapshot registry is reverted to undo all the uncommitted changes related to the write. Hence all the state will be reverted but the state in the offset manager. Now, another operation may come before the snapshot is retried so your offset would be completely wrong now. It seems to me that there is a risk of trimming the log at a wrong offset due to this.

smjn replied:
will revert to timeline hm

smjn committed:
incorporated comments.

smjn committed:
revert to using timeline hm in offsets manager.

smjn committed:
Merge remote-tracking branch 'apache-kafka/trunk' into KAFKA-18058

smjn committed:
fixed javadocs.

smjn committed:
fixed tests.

smjn committed:
incorporated comments.

AndrewJSchofield commented:
@smjn Please resolve the conflicts.

smjn committed:
Merge remote-tracking branch 'apache-kafka/trunk' into KAFKA-18058

smjn committed:
add last red off method

smjn committed:
some debug logs.

smjn commented:
> @smjn Please resolve the conflicts.

@AndrewJSchofield done

smjn committed:
change repeated return values for last red offset.

dajac commented on core/src/main/scala/kafka/coordinator/group/CoordinatorPartitionWriter.scala:
```diff
@@ -165,4 +165,25 @@ class CoordinatorPartitionWriter(
+        if (result.isEmpty) {
+          responseFuture.completeExceptionally(new IllegalStateException(s"Delete status $result should have partition $tp."))
+        } else if (result.get.errorCode() != Errors.NONE.code()) {
+          responseFuture.completeExceptionally(Errors.forCode(result.get.errorCode()).exception())
```
nit: You can omit the `()` for `errorCode`, `code` and `exception`.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+        new TopicPartition("random-topic", 0) -> new DeleteRecordsPartitionResult()
+          .setErrorCode(Errors.NOT_LEADER_OR_FOLLOWER.code())
+      ))
+    })
```
nit: `.thenAnswer(_ => {` -> `.thenAnswer { _ => `

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+    partitionRecordWriter.deleteRecords(
+      new TopicPartition("random-topic", 0),
+      10L
+    ).whenComplete((_, exp) => {
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+      ArgumentMatchers.any(),
+      callbackCapture.capture(),
+      ArgumentMatchers.eq(true)
+    )).thenAnswer(_ => {
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+    partitionRecordWriter.deleteRecords(
+      new TopicPartition("random-topic", 0),
+      10L
+    ).whenComplete((_, exp) => {
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+      ArgumentMatchers.any(),
+      callbackCapture.capture(),
+      ArgumentMatchers.eq(true)
+    )).thenAnswer(_ => {
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+      ))
+    })
+
+    partitionRecordWriter.deleteRecords(
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+    )).thenAnswer(_ => {
+      callbackCapture.getValue.apply(Map(
+        new TopicPartition("random-topic", 0) -> new DeleteRecordsPartitionResult()
+          .setErrorCode(Errors.NOT_LEADER_OR_FOLLOWER.code())
```
nit: You can omit the `()` after `code`.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+    )).thenAnswer(_ => {
+      callbackCapture.getValue.apply(Map(
+        new TopicPartition("random-topic", 0) -> new DeleteRecordsPartitionResult()
+          .setErrorCode(Errors.NONE.code())
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -2852,10 +2853,10 @@ class ReplicaManagerTest {
-    topicPartition: TopicPartition,
-    numOfRecords: Int
-  ): AtomicReference[PartitionResponse] = {
+                                  replicaManager: ReplicaManager,
```
Let's revert all those changes.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3205,15 +3206,15 @@ class ReplicaManagerTest {
-    isolationLevel: IsolationLevel = IsolationLevel.READ_UNCOMMITTED,
-    clientMetadata: Option[ClientMetadata] = None,
-  ): CallbackResult[FetchPartitionData] = {
+                                        replicaManager: ReplicaManager,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3233,14 +3234,14 @@ class ReplicaManagerTest {
-    minBytes: Int = 1,
-    maxBytes: Int = 1024 * 1024,
-  ): CallbackResult[FetchPartitionData] = {
+                                        replicaManager: ReplicaManager,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3255,16 +3256,16 @@ class ReplicaManagerTest {
-    clientMetadata: Option[ClientMetadata],
-    maxWaitMs: Long
-  ): CallbackResult[FetchPartitionData] = {
+                              replicaManager: ReplicaManager,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3289,18 +3290,18 @@ class ReplicaManagerTest {
-    isolation: FetchIsolation = FetchIsolation.LOG_END,
-    clientMetadata: Option[ClientMetadata] = None
-  ): Unit = {
+                               replicaManager: ReplicaManager,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3352,22 +3353,22 @@ class ReplicaManagerTest {
-    buildRemoteLogAuxState: Boolean = false,
-    remoteFetchQuotaExceeded: Option[Boolean] = None
-  ): ReplicaManager = {
+                                                        timer: MockTimer,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3466,11 +3467,11 @@ class ReplicaManagerTest {
-        threadNamePrefix: Option[String],
-        quotaManager: ReplicationQuotaManager
-      ): ReplicaFetcherManager = {
+                                                          metrics: Metrics,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3521,9 +3522,9 @@ class ReplicaManagerTest {
-        quotaManager: ReplicationQuotaManager,
-        brokerTopicStats: BrokerTopicStats
-      ): ReplicaAlterLogDirsManager = {
+                                                     quotaManager: ReplicationQuotaManager,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -4934,16 +4935,16 @@ class ReplicaManagerTest {
-    controllerEpoch: Int = 0,
-    version: Short = LeaderAndIsrRequestData.HIGHEST_SUPPORTED_VERSION
-  ): LeaderAndIsrRequest = {
+                                       topicId: Uuid,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6493,9 +6494,9 @@ class ReplicaManagerTest {
-    replicaManager: ReplicaManager,
-    topicIdPartition: TopicIdPartition
-  ): Unit = {
+                                       replicaManager: ReplicaManager,
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6509,9 +6510,9 @@ class ReplicaManagerTest {
-    replicaManager: ReplicaManager,
-    topicIdPartition: TopicIdPartition
-  ): Unit = {
+                                replicaManager: ReplicaManager,
```
ditto.

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +249,76 @@ public void startup(
+            @Override
+            public void run() {
+                List<CompletableFuture<Void>> futures = new ArrayList<>();
+                runtime.activeTopicPartitions().forEach(tp -> {
+                    futures.add(performRecordPruning(tp));
+                });
```
nit: You could omit the `{}` in this case.

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +249,76 @@ public void startup(
+
+                // None of the futures will complete exceptionally.
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{}))
+                    .thenRun(() -> {
+                        // perpetual recursion
+                        setupRecordPruning();
+                    });
```
You could omit them here too. I also wonder whether we should re-reschedule in case of failure too. `thenRun` would only reschedule if the future succeeds. It would make it more robust to unexpected failures.

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,136 @@
+    // updateState might not be committed yet. In case of retry, these offsets would
+    // be invalidated via the snapshot registry. Hence, using timeline hashmaps
+    // the values would automatically revert in accordance with the last committed offset.
+    private final TimelineHashMap<String, Long> minOffset;
+    private final TimelineHashMap<String, Long> redundantOffset;
```
Have you considered using TimelineObject for those two?

dajac reviewed (commented):
@smjn Thanks for the update. I left some more comments.

smjn committed:
incorporated comments.

smjn committed:
revert changes

dajac commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +249,87 @@ public void startup(
+
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{}))
+                    .whenComplete((res, exp) -> {
+                        if (exp != null) {
+                            log.error("Received error in share state topic prune, stopping job.", exp);
+                            return;
+                        }
```
Are you sure about this? Don't you want to retry after the next interval?

smjn replied:
@dajac 
I think I misunderstood:
```
I also wonder whether we should re-reschedule in case of failure too. thenRun would only reschedule if the future succeeds. It would make it more robust to unexpected failures.
```

`performRecordPruning` which returns the futures never completes the futures exceptionally. I thought you meant that we should not reschedule in case of failures and added completeExceptionally to some cases to stop re-schduling.

Will rectify.

smjn committed:
reschedule even on errors.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,136 @@
+     * Method updates internal state with the supplied offset for the provided
+     * share partition key. It then calculates the minimum offset, if possible,
+     * below which all offsets are redundant. This value is then returned as
+     * an optional.
```
> This value is then returned as an optional.

Hmm, this method has no return value. Ditto for the comment below "The value returned is exclusive".

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,136 @@
+    // be invalidated via the snapshot registry. Hence, using timeline hashmaps
+    // the values would automatically revert in accordance with the last committed offset.
+    private final TimelineLong minOffset;
+    private final TimelineLong redundantOffset;
```
Hmm, `minOffset` is defined as the smallest necessary offset. This seems to to be the same as `redundantOffset`?

smjn replied:
Part of https://github.com/apache/kafka/pull/18014#discussion_r1876665274

junrao replied:
Ok. So `redundantOffset` is the same as `minOffset` but is reset after it's exposed to the caller. It would be useful to document this. Also, I am wondering if it's clearer to maintain just minOffset and a flag that indicates whether it has been exposed or not.

smjn replied:
We could do that

smjn replied:
@junrao incorporated change

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java:
```diff
@@ -378,6 +393,18 @@ public CoordinatorResult<ReadShareGroupStateResponseData, CoordinatorRecord> rea
 
+    /**
+     * Method which returns the last known redundant offset from the partition
+     * lead my this shard.
```
lead my => led by

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,136 @@
+        // reducing efficiency hence, once valid redundant value is
+        // returned lets set it to 0 so next time onwards
+        // empty Optional is returned.
+        redundantOffset.set(0L);
```
Hmm, should we confirm that the subsequent `deleteRecord` call is successful before resetting `redundantOffset`?

smjn replied:
That is not the responsibility of this class. 

The call to this method is from a periodic timer task which keeps querying the method and if a valid value is found - issue the `ReplicaManager.deleteRecords` call. Since the redundant offset will monotonically increase, even if the `deleteRecord` fails for some reason for a lower offset - a higher redundant offset will result in deletion later on.

This value is being set to 0 here to prevent repeated calls to `ReplicaManager.deleteRecords` with the same offset which will not be made if an empty optional (as a consequence of redundant offset being 0) is returned.
This was also the reason for introducing the redundant offset variable as we cannot reset the minimum offset.

smjn replied:
@junrao

junrao reviewed (commented):
@smjn : Thanks for the PR. Made a pass of non-testing files. Added a few comments.

smjn committed:
inc review comments.

smjn commented:
> @smjn : Thanks for the PR. Made a pass of non-testing files. Added a few comments.

@junrao : Thanks for the review.  Incorporated comments, posted response to a some questions

smjn committed:
fixed javadoc

smjn committed:
replaced timeline object with atomic boolean

smjn committed:
updated comments

smjn committed:
Revert "updated comments"

This reverts commit 78d8c49aeafaa61103a3907a73af53814419d065.

smjn committed:
updated docs

junrao commented on core/src/test/scala/unit/kafka/coordinator/group/CoordinatorPartitionWriterTest.scala:
```diff
@@ -238,4 +239,86 @@ class CoordinatorPartitionWriterTest {
+      new TopicPartition("random-topic", 0),
+      10L
+    ).whenComplete { (_, exp) =>
+      assertTrue(exp.isInstanceOf[IllegalStateException])
```
If a topic doesn't exist, we should get an unknown topic exception.

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6660,6 +6661,86 @@ class ReplicaManagerTest {
+    partition.createLogIfNotExists(isNew = false, isFutureReplica = false,
+      new LazyOffsetCheckpoints(rm.highWatermarkCheckpoints.asJava), None)
+
+    rm.becomeLeaderOrFollower(0, new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion, 0, 0, brokerEpoch,
```
`becomeLeaderOrFollower` is only used in ZK based controller, which won't be supported in 4.0. We need to use the cod path for KRaft based controller.

junrao commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManagerTest.java:
```diff
@@ -0,0 +1,209 @@
+                List.of(
+                    ShareOffsetTestHolder.TestTuple.instance(KEY1, 10L, Optional.empty()),
+                    ShareOffsetTestHolder.TestTuple.instance(KEY4, 11L, Optional.empty()),
+                    ShareOffsetTestHolder.TestTuple.instance(KEY2, 13L, Optional.empty())
```
This test result is counter intuitive. I'd expect `lastRedundantOffset` for all three to be 10L since we should be able to truncate the log at offset 10.

smjn replied:
Small efficiency gain here as the offsets will be applied in increasing order (partition records auto inc offset) to the `updateState` method. If the case is such that 10L is the smallest one - it means there are no offsets smaller than that present in the topic partition, hence returning 10L will be of no consequence and save an extra `deleteRecords` call.

In the algorithm we have chosen to get at least 2 offsets for any key before exposing the redundant offset. 

Also since we are exposing the offset only once and then setting the boolean flag, the 2nd and 3rd line should be `empty`.

junrao commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManagerTest.java:
```diff
@@ -0,0 +1,209 @@
+            ),
+
+            new ShareOffsetTestHolder(
+                "redundant state cold partition",
```
What does cold partition mean?

smjn replied:
infrequently written

junrao commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorServiceTest.java:
```diff
@@ -274,6 +290,7 @@ public void testReadStateSuccess() throws ExecutionException, InterruptedExcepti
                 )))
             );
 
+
```
extra new line

junrao commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorServiceTest.java:
```diff
@@ -673,4 +706,348 @@ public void testPartitionFor() {
+    @Test
+    public void testRecordPruningTaskPeriodicityWithAllSuccess() throws Exception {
+        CoordinatorRuntime<ShareCoordinatorShard, CoordinatorRecord> runtime = mockRuntime();
+        org.apache.kafka.server.util.MockTime time = new org.apache.kafka.server.util.MockTime();
```
Could we import MockTime?

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,143 @@
+     * After returning the value once, the redundant offset is reset.
+     * @return Optional of type Long representing the offset or empty for invalid offset values
+     */
+    public Optional<Long> lastRedundantOffset() {
```
This method is ok, but is very customized for the usage in the current only caller. If there is another caller, it can unexpectedly break the existing caller. A better api is probably to always expose the lastRedundantOffset and let the caller handle the case when the same value is returned.

junrao reviewed (commented):
@smjn : Thanks for the updated PR. Made a pass of all files. A few more comments.

smjn committed:
incorporated comments.

smjn commented:
> @smjn : Thanks for the updated PR. Made a pass of all files. A few more comments.

@junrao Thanks again for the review, incorporated all comments.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,122 @@
+    public ShareCoordinatorOffsetsManager(SnapshotRegistry snapshotRegistry) {
+        Objects.requireNonNull(snapshotRegistry);
+        offsets = new TimelineHashMap<>(snapshotRegistry, 0);
+        minOffset = new TimelineLong(snapshotRegistry);
```
minOffset => lastRedundantOffset ?

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java:
```diff
@@ -0,0 +1,122 @@
+        minOffset.set(Math.min(minOffset.get(), offset));
+        offsets.put(key, offset);
+
+        Optional<Long> deleteTillOffset = findRedundantOffset();
```
deleteTillOffset => redundantOffset ?

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +251,96 @@ public void startup(
+            if (result.isPresent()) {
+                Long off = result.get();
+                // Guard and optimization.
+                if (off == Long.MAX_VALUE || off <= 0) {
```
This test seems redundant since `ShareCoordinatorOffsetsManager.lastRedundantOffset` does that already.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +251,96 @@ public void startup(
 
         log.info("Starting up.");
         numPartitions = shareGroupTopicPartitionCount.getAsInt();
+        Map<TopicPartition, Long> offsets = new ConcurrentHashMap<>();
```
1. offsets => lastPrunedOffsets? 
2. Would it be better to make that an instance val so that we don't have to pass it around?
3. Should we remove entries when `onResignation` is called?

junrao commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -6660,6 +6661,61 @@ class ReplicaManagerTest {
   }
 
+  @Test
+  def testDeleteRecordsInternalTopicDeleteDisallowed(): Unit = {
```
There are lots of existing usage of `rm.becomeLeaderOrFollower`. It would be useful to clean them up in a followup jira.

smjn replied:
perhaps

junrao reviewed (commented):
@smjn : Thanks for the updated PR. A few more comments.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +251,96 @@ public void startup(
+    private CompletableFuture<Void> performRecordPruning(TopicPartition tp, Map<TopicPartition, Long> offsets) {
+        // This future will always be completed normally, exception or not.
+        CompletableFuture<Void> fut = new CompletableFuture<>();
+        runtime.scheduleWriteOperation(
```
This call doesn't do any writes in runtime. Should we use scheduleReadOperation? Similarly, I also don't understand why `readState` calls `runtime.scheduleWriteOperation`.

smjn replied:
No, the write operation used here is for the write consistency offered by the method.

The `ShareCoordinatorShard.replay` calls `offsetsManager.updateState` with various last written offset values. The `replay` method itself is called when other write RPCs produce records. However, it does not mean the offset set in replay has been committed.

Now, the coordinator enqueues the write operations in a queue and guarantees that when the `scheduleWriteOperation` completes, the records it generated have been replicated, even those which were written before it.

The framework however, gives no consistency guarantees between write and read operations. Consider, a write op writing an offset into the offset manager. We only know that this offset is written but not replicated. A subsequent read could give us back the same offset but still there is no guarantee that this offset has been replicated. It is only when the next write operation completes, do we have a guarantee that the previous offset has been committed. 

This was extensively discussed with the coordinator framework owner @dajac and we arrived at this solution.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +251,96 @@ public void startup(
+                        fut.complete(null);
+                        // Update offsets map as we do not want to
+                        // issue repeated deleted
+                        offsets.put(tp, off);
```
Since this is called from the purgatory thread, it's possible when this call occurs, the partition leader has already resigned and we need to handle that accordingly.

smjn replied:
I don't foresee any consistency issues with that - at max a repeated delete call might be made which is acceptable since the frequency of these calls is very low (in minutes).
Whoever the leader, the partition offsets will not change.

smjn committed:
incorporated comments.

smjn commented:
> @smjn : Thanks for the updated PR. A few more comments.

@junrao Thanks for the review, incorporated few changes. Replied to some suggestions.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -543,6 +629,7 @@ public void onElection(int partitionIndex, int partitionLeaderEpoch) {
     @Override
     public void onResignation(int partitionIndex, OptionalInt partitionLeaderEpoch) {
         throwIfNotActive();
+        lastPrunedOffsets.clear();
```
We should only clear the entry for partitionIndex, right?

smjn replied:
Yes, makes sense

junrao reviewed (commented):
@smjn : Thanks for the updated PR. Just one more comment.

smjn committed:
remove part index, instead of clearing memo.

smjn commented:
> @smjn : Thanks for the updated PR. Just one more comment.

@junrao Thanks for the review, incorporated change

junrao reviewed (approved):
@smjn : Thanks for the updated PR. LGTM. @AndrewJSchofield and @dajac, any other comments from you?

dajac commented:
@junrao No. All good to me. Thanks!

AndrewJSchofield commented:
@junrao Thanks for the review. I'll take another look this evening and merge once I'm happy.

AndrewJSchofield commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +253,82 @@ public void startup(
+            if (result.isPresent()) {
+                Long off = result.get();
+
+                if (lastPrunedOffsets.containsKey(tp) && Objects.equals(lastPrunedOffsets.get(tp), off)) {
```
I suggest `lastPrunedOffsets.get(tp).longValue() == off`. Using the generic object equality method seems odd for just a pair of longs.

AndrewJSchofield commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -194,12 +202,17 @@ public ShareCoordinatorService(
-        Time time) {
+        Time time,
+        Timer timer,
+        PartitionWriter writer) {
```
nit: New line please so the arguments and the method body do not run into each other.

AndrewJSchofield commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +253,82 @@ public void startup(
     }
 
+    private void setupRecordPruning() {
+        log.info("Scheduling share state topic prune job.");
```
`share-group state topic` is what we use in most places.

AndrewJSchofield commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +253,82 @@ public void startup(
+                CompletableFuture.allOf(futures.toArray(new CompletableFuture[]{}))
+                    .whenComplete((res, exp) -> {
+                        if (exp != null) {
+                            log.error("Received error in share state topic prune.", exp);
```
`share-group state topic`

AndrewJSchofield commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +253,82 @@ public void startup(
+                Long off = result.get();
+
+                if (lastPrunedOffsets.containsKey(tp) && Objects.equals(lastPrunedOffsets.get(tp), off)) {
+                    log.debug("{} already pruned at offset {}", tp, off);
```
You've used `till` in most places, so I'd replace the `at` here too.

AndrewJSchofield commented on share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorConfigTest.java:
```diff
@@ -50,6 +50,7 @@ private static Map<String, String> testConfigMapRaw() {
         configs.put(ShareCoordinatorConfig.LOAD_BUFFER_SIZE_CONFIG, "555");
         configs.put(ShareCoordinatorConfig.APPEND_LINGER_MS_CONFIG, "10");
         configs.put(ShareCoordinatorConfig.STATE_TOPIC_COMPRESSION_CODEC_CONFIG, String.valueOf(CompressionType.NONE.id));
+        configs.put(ShareCoordinatorConfig.STATE_TOPIC_PRUNE_INTERVAL_MS_CONFIG, "30000");  // 30 seconds
```
This class doesn't contain any tests, so calling it XYZTest is peculiar. Please rename to `ShareCoordinatorTestUtils` or similar.

AndrewJSchofield reviewed (changes requested):
A handful of final review comments. I've been running with the pruning enabled and it's happily deleting records as intended.

smjn committed:
incorporated comments.

smjn commented:
> A handful of final review comments. I've been running with the pruning enabled and it's happily deleting records as intended.

@AndrewJSchofield Thanks for the review, incorporated comments.

junrao commented on share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java:
```diff
@@ -240,9 +253,82 @@ public void startup(
+                        }
+                        fut.complete(null);
+                        // Best effort prevention of issuing duplicate delete calls.
+                        lastPrunedOffsets.put(tp, off);
```
This approach is ok, but it breaks the pattern in CoordinatorRuntime that all internal states are updated by CoordinatorRuntime threads since lastPrunedOffsets is now updated by the request io threads. We probably could follow the approach of how CoordinatorRuntime waits for a record to be replicated. It appends the record to local log and registers a `PartitionListener` for `onHighWatermarkUpdated()`. Once a new HWM is received, `CoordinatorRuntime.onHighWatermarkUpdated` enqueues a `CoordinatorInternalEvent`, the processing of which will trigger the update of the internal state. We could follow a similar approach for `LowWaterMark`. This can be done in a follow up jira.

smjn replied:
While my initial approach was to add functionality in `CoordinatorRuntime` - the requirement wasn't general enough to modify the runtime.

AndrewJSchofield replied:
I'm sure this is not the final state of this code, but it's definitely a fine starting point.
